lib/erlef/jobs.ex

```diff
@@ -0,0 +1,354 @@
+          country: String.t() | nil,
+          remote: boolean(),
+          website: URI.t()
+        }
+
+  @doc """
+  Retrieves all job posts which are approved and not expired.
+
+  Sposored posts are in front of the list.
+  """
+  @spec list_posts() :: [Post.t()]
+  def list_posts() do
+    Post.where_approved()
+    |> Post.where_fresh()
+    |> Post.order_by_sponsor_owner_asc()
+    |> Repo.all()
+  end
+
+  @doc """
+  Retrieves all job posts which are not approved.
+  """
+  @spec list_unapproved_posts() :: [Post.t()]
+  def list_unapproved_posts() do
+    Post.from()
+    |> Post.where_unapproved()
+    |> Repo.all()
+  end
+
+  @doc """
+  Retrieves all job posts by an author with the provided id.
+  """
+  @spec list_posts_by_author_id(term()) :: [Post.t()]
+  def list_posts_by_author_id(author_id) do
+    author_id
+    |> Post.where_author_id()
+    |> Repo.all()
+  end
+
+  @doc """
+  Retrieves a job post by its id.
+  """
+  @spec get_post!(term()) :: Post.t()
+  def get_post!(id) do
+    id
+    |> Post.where_id()
+    |> Repo.one!()
+  end
+
+  @doc """
+  Creates a job post.
+
+  A job post can be created by a sponsor-associated member if the sponsor hasn't
+  yet reached the posts quota. If the member is not associated to a sponsor, the
+  rule is applied to the member itself.
+  """
+  @spec create_post(Member.t(), create_post_params()) ::
+          {:ok, Post.t()} | {:error, __MODULE__.Error.t()}
+  def create_post(%Member{} = member, attrs \\ %{}) do
+    Repo.transact(fn ->
+      with {:authz, true} <- {:authz, can_create_post?(member)},
+           {:ok, post} <- do_insert_post(member, attrs),
+           created_by = post.created_by,
+           :ok <- update_post_history(:insert, created_by, post),
+           :ok <- send_post_submission_notifications(created_by) do
+        {:ok, post}
+      else
+        {:authz, false} ->
+          {:error, Error.exception(:post_quota_reached)}
+
+        {:error, %Changeset{} = cs} ->
+          {:error, Error.exception({:changeset, cs})}
+      end
+    end)
+  end
+
+  @spec do_insert_post(Member.t(), create_post_params()) ::
+          {:ok, Post.t()} | {:error, Changeset.t()}
+  defp do_insert_post(member, attrs) do
+    result =
+      member
+      |> Ecto.build_assoc(:posts)
+      |> Post.changeset(attrs)
+      |> Repo.insert()
+
+    case result do
+      {:ok, post} ->
+        {:ok, Map.put(post, :updated_by, post.created_by)}
+
+      error ->
+        error
+    end
+  end
+
+  @doc """
+  Approves a job post.
+
+  The job post can be approved only by a member who is an admin.
+  """
+  @spec approve_post(Member.t(), Post.t()) :: {:ok, Post.t()} | {:error, :unauthorized}
+  def approve_post(%Member{is_app_admin: true} = member, %Post{} = post) do
+    with {:ok, %Post{} = post} <- update_post(member, post, %{approved: true}),
+         {:ok, _} <- send_post_approval_notification(post.created_by, post.title) do
+      {:ok, post}
+    else
```

Comment:
We shouldn't need the `else` clause here, `with` will just return the error, which is the best `with` :)